thesis_text/thesis_text.md

```diff
@@ -0,0 +1,551 @@
+
+> The Common Language Infrastructure (CLI) provides a specification for executable code and the execution environment (the Virtual Execution System) in which it runs. 
+
+.NET languages (like C#) are compiled into "managed code"[^7] - instead of targeting native processor instruction sets, they target the CLI's execution environment.
+
+Using the definitions of the standard, BACIL is actually a Virtual Execution System (VES):
+
+> The VES is responsible for loading and running programs written for the CLI. It provides the services needed to execute managed code and data, using the metadata to connect separately generated modules together at runtime (late binding).
+
+.NET Framework's VES is called the Common Language Runtime (CLR) and in .NET Core, it's known as CoreCLR.
+
+> To a large extent, the purpose of the VES is to provide the support required to execute the [Common Intermediate Language (CIL)] instruction set.
+
+The CIL, historically also called Microsoft Intermediate Language (MSIL) or simply Intermediate Language (IL), is the instruction set used by the CLI. Interpreting (a subset of) this instruction set was the main goal of this work.
+
+Another large part of the framework is the standard libraries - the base class library, which has to be supported by all implementations of the CLI, consists of 2370 members over 207 classes. As the focus of the work was on the core interpreter, we largely ignored this part of the standard and deferred to other standard library implementations where possible.
+
+## Truffle and Graal
+
+To implement a high-performance CLI runtime, we alleviate the 
+ [Truffle language implementation framework](https://www.graalvm.org/graalvm-as-a-platform/language-implementation-framework/) (henceforth "Truffle") and the [GraalVM Compiler](https://www.graalvm.org/22.1/docs/introduction). These two components are tightly coupled together and we'll mostly be referring to them that way. 
+ 
+The Graal Compiler is a general high-performance just-in-time compiler for Java bytecode that is itself written in Java. It is state-of-the-art when it comes to optimization algorithms - according to official documentation, "the compiler in GraalVM Enterprise includes 62 optimization phases, of which 27 are patented". 
+
+Truffle on the other hand is a framework for implementing languages that will run on Graal. From the outside, it behaves like a compiler: its job is to take guest language code and convert it to the VM's language. Unlike a hand-crafted compiler, Truffle takes an interpreter of the guest language as its input and uses [Partial evaluation](#partial-evaluation) to do the compilation, performing a so-called "first Futamura projection".
```

Comment:
will be compiled by Graal? or will run on GraalVM? Nekdy se rozlisuje Truffle framework nebo API (implementacni framework pro autory interpretru) a Truffle compiler == Graal compiler + compiler pluginy pro partial evaluation, ktere rozumneji tomu Truffle API